lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -1629,6 +1639,146 @@ struct ConvertLayoutOpConversion
+    // thus the adjacent elems may not belong to a same word. This could be
+    // improved if we update the elements order by emitIndicesForBlockedLayout()
+    SmallVector<unsigned> wordsInEachRep(2);
+    wordsInEachRep[0] = inOrd[0] == 0
+                            ? srcBlockedLayout.getSizePerThread()[0] / minVec
+                            : srcBlockedLayout.getSizePerThread()[0];
+    wordsInEachRep[1] = inOrd[0] == 0
+                            ? srcBlockedLayout.getSizePerThread()[1]
+                            : srcBlockedLayout.getSizePerThread()[1] / minVec;
+    Value outVecVal = createIndexConst(rewriter, loc, outVec);
+    Value minVecVal = createIndexConst(rewriter, loc, minVec);
+    Value smemBase = getSharedMemoryBase(loc, rewriter, dst);
+    auto elemPtrTy =
+        LLVM::LLVMPointerType::get(getTypeConverter()->convertType(elemTy), 3);
+    smemBase = rewriter.create<LLVM::BitcastOp>(loc, elemPtrTy, smemBase);
+    unsigned numWordsEachRep = product<unsigned>(wordsInEachRep);
+    SmallVector<Value> wordVecs(numWordsEachRep);
+    for (unsigned i = 0; i < numElems; ++i) {
+      if (i % srcAccumSizeInThreads == 0) {
+        // start of a replication
+        for (unsigned w = 0; w < numWordsEachRep; ++w) {
+          wordVecs[w] = rewriter.create<LLVM::UndefOp>(loc, wordTy);
+        }
+      }
+      unsigned linearIdxInNanoTile = i % srcAccumSizeInThreads;
+      auto multiDimIdxInNanoTile = getMultiDimIndex<unsigned>(
+          i % srcAccumSizeInThreads, srcBlockedLayout.getSizePerThread());
+      multiDimIdxInNanoTile[inOrd[0]] /= minVec;
+      unsigned pos = multiDimIdxInNanoTile[inOrd[0]] % minVec;
+      unsigned wordVecIdx =
+          getLinearIndex<unsigned>(multiDimIdxInNanoTile, wordsInEachRep);
+      wordVecs[wordVecIdx] = rewriter.create<LLVM::InsertElementOp>(
+          loc, wordTy, wordVecs[wordVecIdx], inVals[i],
+          createIndexConst(rewriter, loc, pos));
+
+      if (i % srcAccumSizeInThreads == srcAccumSizeInThreads - 1) {
+        // end of replication, store the vectors into shared memory
+        unsigned linearRepIdx = i / srcAccumSizeInThreads;
+        auto multiDimRepIdx = getMultiDimIndex<unsigned>(linearRepIdx, reps);
+        for (unsigned linearWordIdx = 0; linearWordIdx < numWordsEachRep;
+             ++linearWordIdx) {
+          // step 1: recover the multidim_index from the index of input_elements
+          // row = multiDimOffsetFirstElem[0] +
+          //       multiDimRepIdx[0] * shapePerCTA[0] + multiDimWordIdx[0] *
+          //       (inOrd[0] == 0) ? minVec : 1
+          // col = multiDimOffsetFirstElem[1] +
+          //       multiDimRepIdx[1] * shapePerCTA[1] + multiDimWordIdx[1] *
+          //       (inOrd[0] == 1) ? minVec : 1
+          auto multiDimWordIdx =
+              getMultiDimIndex<unsigned>(linearWordIdx, wordsInEachRep);
+          SmallVector<Value> multiDimIdx(2);
+          multiDimIdx[0] = rewriter.create<LLVM::AddOp>(
```

Comment:
Could we have an `add` macro here that's similar to https://github.com/openai/triton/blob/08093793240ebd7a013d8c195d2f3594467334d1/lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp#L75 ?

Reply:
ditto